Plugins/Flow.Launcher.Plugin.Program/Views/Models/ProgramSource.cs

```diff
@@ -1,5 +1,91 @@
-﻿
+﻿using System;
+using System.IO;
+using Flow.Launcher.Plugin.Program.Programs;
+
 namespace Flow.Launcher.Plugin.Program.Views.Models
 {
-    public class ProgramSource : Settings.ProgramSource { }
+    /// <summary>
+    /// Contains user added folder location contents as well as all user disabled applications
+    /// </summary>
+    /// <remarks>
+    /// <para>Win32 class applications set UniqueIdentifier using their full file path</para>
+    /// <para>UWP class applications set UniqueIdentifier using their Application User Model ID</para>
+    /// <para>Custom user added program sources set UniqueIdentifier using their location</para>
+    /// </remarks>
+    public class ProgramSource
+    {
+        private string name;
+
+        public string Location { get; set; }
+        public string Name { get => name ?? new DirectoryInfo(Location).Name; set => name = value; }
+        public bool Enabled { get; set; } = true;
+        private string uid { get; set; }
+
+        /// <summary>
+        /// Guranteed lowercase.
+        /// </summary>
+        public string UniqueIdentifier { get => uid; set => uid = value.ToLowerInvariant(); }
+
+        public ProgramSource() { }  // only for json deserialization
+
+        /// <summary>
+        /// Add source by location
+        /// </summary>
+        /// <param name="location">location of program source</param>
+        /// <param name="enabled">enabled</param>
+        public ProgramSource(string location, bool enabled=true)
+        {
+            Location = location;
+            Enabled = enabled;
+            UniqueIdentifier = location;
+        }
+
+        public ProgramSource(ProgramSource source)
+        {
+            Location = source.Location;
+            Name = source.Name;
+            Enabled = source.Enabled;
+            UniqueIdentifier = source.UniqueIdentifier;
+        }
+
+        public ProgramSource(IProgram source)
+        {
+            Location = source.Location;
+            Name = source.Name;
+            Enabled = source.Enabled;
+            UniqueIdentifier = source.UniqueIdentifier;
+        }
+
+        public override bool Equals(object obj)
+        {
+            return obj is ProgramSource other && other.UniqueIdentifier == this.UniqueIdentifier;
+        }
+
+        public bool Equals(IProgram program)
+        {
+            return program != null && program.UniqueIdentifier == this.UniqueIdentifier;
+        }
+
+        public override int GetHashCode()
+        {
+            return HashCode.Combine(uid);
+        }
+    }
+
+    public class DisabledProgramSource : ProgramSource
```

Comment:
I don't understand why we need an inheritance here....It only changes the mutable Property.....Could you explain your idea here?

Further, the property `Enabled` is public, so this subclass doesn't restrict people to change `Enabled` to `true`.

Reply:
I don't quite understand either. It's not my design though, I just keep it. Maybe the original purpose was not to mess up the disabled and the enabled.

Reply:
After some research I think `DisabledProgramSource` is only for a disabled SINGLE program. If you disable a single program from context menu, a new instance is added to the list.

Reply:
Already removed `DisabledProgramSource`. Using `ProgramSource` instead.